kubernetes/provider.go

```diff
@@ -192,55 +191,100 @@ type KubeClientsets struct {
 
-	var cfg *restclient.Config
-	var err error
-	if d.Get("load_config_file").(bool) {
-		// Config file loading
-		cfg, err = tryLoadingConfigFile(d)
-	}
-
+	// Config initialization
+	cfg, err := initializeConfiguration(d)
 	if err != nil {
 		return nil, err
 	}
 	if cfg == nil {
-		// Attempt to load in-cluster config
-		cfg, err = restclient.InClusterConfig()
-		if err != nil {
-			// Fallback to standard config if we are not running inside a cluster
-			if err == restclient.ErrNotInCluster {
-				cfg = &restclient.Config{}
-			} else {
-				return nil, fmt.Errorf("Failed to configure: %s", err)
+		return nil, fmt.Errorf("Failed to initialize config")
+	}
+
+	cfg.UserAgent = fmt.Sprintf("HashiCorp/1.0 Terraform/%s", terraformVersion)
+
+	if logging.IsDebugOrHigher() {
+		log.Printf("[DEBUG] Enabling HTTP requests/responses tracing")
+		cfg.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
+			return logging.NewTransport("Kubernetes", rt)
+		}
+	}
+
+	k, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to configure: %s", err)
+	}
+
+	a, err := aggregator.NewForConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to configure: %s", err)
+	}
+
+	return &KubeClientsets{k, a}, nil
+}
+
+func initializeConfiguration(d *schema.ResourceData) (*restclient.Config, error) {
+	overrides := &clientcmd.ConfigOverrides{}
+	loader := &clientcmd.ClientConfigLoadingRules{}
+
+	if d.Get("load_config_file").(bool) {
+		if configPath, ok := d.GetOk("config_path"); ok && configPath.(string) != "" {
+			path, err := homedir.Expand(configPath.(string))
+			if err != nil {
+				return nil, err
+			}
+			loader.ExplicitPath = path
+
+			ctxSuffix := "; default context"
+
+			ctx, ctxOk := d.GetOk("config_context")
+			authInfo, authInfoOk := d.GetOk("config_context_auth_info")
+			cluster, clusterOk := d.GetOk("config_context_cluster")
+			if ctxOk || authInfoOk || clusterOk {
+				ctxSuffix = "; overriden context"
+				if ctxOk {
+					overrides.CurrentContext = ctx.(string)
+					ctxSuffix += fmt.Sprintf("; config ctx: %s", overrides.CurrentContext)
+					log.Printf("[DEBUG] Using custom current context: %q", overrides.CurrentContext)
+				}
+
+				overrides.Context = clientcmdapi.Context{}
+				if authInfoOk {
+					overrides.Context.AuthInfo = authInfo.(string)
+					ctxSuffix += fmt.Sprintf("; auth_info: %s", overrides.Context.AuthInfo)
+				}
+				if clusterOk {
+					overrides.Context.Cluster = cluster.(string)
+					ctxSuffix += fmt.Sprintf("; cluster: %s", overrides.Context.Cluster)
+				}
+				log.Printf("[DEBUG] Using overidden context: %#v", overrides.Context)
 			}
 		}
 	}
 
 	// Overriding with static configuration
-	cfg.UserAgent = fmt.Sprintf("HashiCorp/1.0 Terraform/%s", terraformVersion)
-
 	if v, ok := d.GetOk("host"); ok {
-		cfg.Host = v.(string)
+		overrides.ClusterInfo.Server = v.(string)
```

Comment:
This change causes a regression when passing a bare IP address instead of a URL, e.g. when using this:
```
provider "kubernetes" {
  version = "1.10.1-dev3"

  load_config_file = false
  host             = data.google_container_cluster.cluster0.endpoint
  token            = data.google_client_config.default.access_token
  cluster_ca_certificate = base64decode(
    data.google_container_cluster.cluster0.master_auth[0].cluster_ca_certificate,
  )
}
```

I believe that's because `overrides.ClusterInfo.Server` always expects a URL: https://github.com/kubernetes/client-go/blob/v12.0.0/tools/clientcmd/api/types.go#L68-L69

While `cfg.Host` is most flexible: https://github.com/kubernetes/client-go/blob/v12.0.0/rest/config.go#L53-L57

Reply:
A work-around is to use this:

```
  host             = "https://${data.google_container_cluster.cluster0.endpoint}:443"
```

But I'm experimenting how to avoid this.

Reply:
Resolved.

Reply:
Not the proper way to do it.

Reply:
Alright, `overrides` are processed too late to be taken into account by `defaultServerUrlFor()`:

- https://github.com/kubernetes/client-go/blob/v12.0.0/tools/clientcmd/client_config.go#L176
- https://github.com/kubernetes/client-go/blob/v12.0.0/rest/config.go#L438-L439
- https://github.com/kubernetes/client-go/blob/v12.0.0/rest/url_utils.go#L85
- https://github.com/kubernetes/client-go/blob/v12.0.0/tools/clientcmd/client_config.go#L189
- https://github.com/kubernetes/client-go/blob/v12.0.0/tools/clientcmd/client_config.go#L211